File: src/python/grpc/grpcio/service.py
"""BenchmarkService gRPC implementation."""

import datetime
import uuid

import grpc
from canonical import (
    CANONICAL_CREATED_AT,
    canonical_email,
    canonical_is_active,
    canonical_name,
    canonical_uuid,
    item_count,
)

# Generated stubs -- available after running generate.py
from generated import benchmark_pb2
from generated import benchmark_pb2_grpc

from cache import get_client as get_redis
from db import check_database, get_connection, release_connection


class BenchmarkServiceServicer(benchmark_pb2_grpc.BenchmarkServiceServicer):
    """Implementation of BenchmarkService."""

    VERSION = "1.0.0"

    def Health(self, request, context):
        """Scenario 1: Health check."""
        from cache import check_cache

        return benchmark_pb2.HealthResponse(
            status="ok",
            version=self.VERSION,
            timestamp=datetime.datetime.utcnow().isoformat() + "Z",
            database=check_database(),
            cache=check_cache(),
        )

    def GetJsonItems(self, request, context):
        """Scenario 2: JSON serialization (1000 items)."""
        # The previous version minted a uuid.uuid4() per item -- 1000 random
        # UUIDs per request -- numbered items from 1 and named them "user_3"
        # at user_3@benchmark.dev. See contracts/rest/canonical-payloads.md.
        limit = item_count(request.limit)
        items = [
            benchmark_pb2.JsonItem(
                id=i,
                uuid=canonical_uuid(i),
                name=canonical_name(i),
                email=canonical_email(i),
                created_at=CANONICAL_CREATED_AT,
                is_active=canonical_is_active(i),
            )
            for i in range(limit)
        ]
        return benchmark_pb2.JsonItemsResponse(
            items=items,
            count=len(items),
            timestamp=datetime.datetime.utcnow().isoformat() + "Z",
        )

    def GetUser(self, request, context):
        """Scenario 3: Simple database query (single row)."""
        conn = get_connection()
        try:
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT id, email, first_name, last_name, age, created_at "
                    "FROM users WHERE id = %s",
                    (request.id,),
                )
                row = cur.fetchone()
                if row is None:
                    context.abort(grpc.StatusCode.NOT_FOUND, f"User {request.id} not found")
                return benchmark_pb2.UserResponse(
                    id=row[0],
                    email=row[1],
                    first_name=row[2],
                    last_name=row[3],
                    age=row[4],
                    created_at=str(row[5]),
                )
        finally:
            release_connection(conn)

    def GetComplexOrders(self, request, context):
        """Scenario 4: Complex database query (JOIN + aggregation)."""
        days = request.days if request.days > 0 else 30
        conn = get_connection()
        try:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    -- Normative SQL, see contracts/rest/canonical-payloads.md. The
                    -- previous query aggregated o.total or o.amount, columns the schema
                    -- does not have; the interval was pasted in as text rather than
                    -- bound, so it was never substituted; the ORDER BY had no tiebreak;
                    -- and there was no LIMIT, so it returned every user rather than the
                    -- 100 the contract fixes -- which changes the payload size and the
                    -- network ceiling of this scenario.
                    SELECT
                        u.id AS user_id,
                        u.first_name || ' ' || u.last_name AS user_name,
                        COUNT(o.id) AS total_orders,
                        COALESCE(SUM(o.total_amount), 0) AS total_value,
                        COALESCE(AVG(o.total_amount), 0) AS average_order_value
                    FROM users u
                    INNER JOIN orders o ON u.id = o.user_id
                        WHERE o.created_at >= NOW() - INTERVAL '1 day' * %s
                    GROUP BY u.id, u.first_name, u.last_name
                    ORDER BY total_orders DESC, u.id
                    LIMIT 100
                    """,
                    (days,),
                )
                rows = cur.fetchall()
                data = []
                for row in rows:
                    data.append(
                        benchmark_pb2.UserOrderStats(
                            user_id=row[0],
                            user_name=row[1],
                            total_orders=row[2],
                            total_value=float(row[3]),
                            average_order_value=float(row[4]),
                        )
                    )
                return benchmark_pb2.ComplexOrdersResponse(
                    period_days=days,
                    total_users=len(data),
                    data=data,
                )
        finally:
            release_connection(conn)

    def GetCacheValue(self, request, context):
        """Scenario 5: Cache hit/miss."""
        client = get_redis()
        cached = client.get(request.key)
        if cached is not None:
            ttl = client.ttl(request.key)
            return benchmark_pb2.CacheResponse(
                key=request.key,
                value=cached,
                cached=True,
                ttl=ttl if ttl >= 0 else 0,
                timestamp=datetime.datetime.utcnow().isoformat() + "Z",
            )

        # Cache miss: generate a value and store it
        value = f"value_for_{request.key}_{uuid.uuid4()}"
        client.setex(request.key, 3600, value)
        return benchmark_pb2.CacheResponse(
            key=request.key,
            value=value,
            cached=False,
            ttl=3600,
            timestamp=datetime.datetime.utcnow().isoformat() + "Z",
        )

File: src/deno/grpc/nice-grpc/service.ts
import * as db from "./db.ts";
import * as cache from "./cache.ts";
import {
  CANONICAL_CREATED_AT,
  canonicalEmail,
  canonicalIsActive,
  canonicalName,
  canonicalUuid,
  itemCount,
} from './canonical.ts';

const VERSION = Deno.env.get("APP_VERSION") || "1.0.0";

/**
 * BenchmarkService implementation for nice-grpc on Deno.
 * nice-grpc service methods receive (request, context) and return the response directly.
 */
export default {
  // Scenario 1: Health check
  async health(request: Record<string, unknown>, _context: unknown) {
    try {
      const [dbStatus, cacheStatus] = await Promise.all([
        db.healthCheck(),
        cache.healthCheck(),
      ]);

      return {
        status: "ok",
        version: VERSION,
        timestamp: new Date().toISOString(),
        database: dbStatus,
        cache: cacheStatus,
      };
    } catch (err) {
      console.error("Health check error:", err);
      return {
        status: "degraded",
        version: VERSION,
        timestamp: new Date().toISOString(),
        database: "unknown",
        cache: "unknown",
      };
    }
  },

  // Scenario 2: JSON serialization (1000 items)
  async getJsonItems(request: Record<string, unknown>, _context: unknown) {
    const limit = (request.limit as number) > 0 ? (request.limit as number) : 1000;
    const items = [];

    for (let i = 0; i < count; i++) {
      items.push({
        id: i,
        uuid: canonicalUuid(i),
        name: canonicalName(i),
        email: canonicalEmail(i),
        created_at: CANONICAL_CREATED_AT,
        is_active: canonicalIsActive(i),
      });
    }

    return {
      items,
      count: items.length,
      timestamp: new Date().toISOString(),
    };
  },

  // Scenario 3: Simple database query
  async getUser(request: Record<string, unknown>, _context: unknown) {
    const id = request.id as number;
    const result = await db.query(
      "SELECT id, email, first_name, last_name, age, created_at FROM users WHERE id = $1",
      [id]
    );

    if (result.rows.length === 0) {
      const err = new Error(`User with id ${id} not found`) as Error & { code: number };
      err.code = 5; // NOT_FOUND
      throw err;
    }

    const user = result.rows[0] as Record<string, unknown>;
    return {
      id: user.id as number,
      email: (user.email as string) || "",
      first_name: (user.first_name as string) || "",
      last_name: (user.last_name as string) || "",
      age: (user.age as number) || 0,
      created_at: user.created_at instanceof Date
        ? (user.created_at as Date).toISOString()
        : String(user.created_at || ""),
    };
  },

  // Scenario 4: Complex database query (JOIN + aggregation)
  async getComplexOrders(request: Record<string, unknown>, _context: unknown) {
    const days = (request.days as number) > 0 ? (request.days as number) : 30;

    const result = await db.query(
      `SELECT
        u.id as user_id,
        u.first_name || ' ' || u.last_name as user_name,
        COUNT(o.id) as total_orders,
        COALESCE(SUM(o.total_amount), 0) as total_value,
        COALESCE(AVG(o.total_amount), 0) as average_order_value
      FROM users u
      LEFT JOIN orders o ON u.id = o.user_id
        AND o.created_at >= NOW() - INTERVAL '${days} days'
      GROUP BY u.id, u.first_name, u.last_name
      HAVING COUNT(o.id) > 0
      ORDER BY total_value DESC
      LIMIT 100`
    );

    const data = result.rows.map((row: Record<string, unknown>) => ({
      user_id: row.user_id as number,
      user_name: (row.user_name as string) || "",
      total_orders: parseInt(row.total_orders as string) || 0,
      total_value: parseFloat(row.total_value as string) || 0,
      average_order_value: parseFloat(row.average_order_value as string) || 0,
    }));

    return {
      period_days: days,
      total_users: data.length,
      data,
    };
  },

  // Scenario 5: Cache hit/miss
  async getCacheValue(request: Record<string, unknown>, _context: unknown) {
    const key = request.key as string;
    const result = await cache.get(key);

    if (result.hit) {
      return {
        key,
        value: result.value,
        cached: true,
        ttl: 300,
        timestamp: new Date().toISOString(),
      };
    }

    // Cache miss - generate a value and store it
    const value = `value-${Date.now()}`;
    await cache.set(key, value, 300);

    return {
      key,
      value,
      cached: false,
      ttl: 300,
      timestamp: new Date().toISOString(),
    };
  },
};
